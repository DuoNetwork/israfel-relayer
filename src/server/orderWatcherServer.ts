import {
	ExchangeContractErrs,
	OrderState,
	OrderStateInvalid,
	OrderStateValid,
	OrderWatcher
} from '0x.js';
import SequenceClient from '../client/SequenceClient';
import * as CST from '../common/constants';
import {
	ILiveOrder,
	IOption,
	IOrderQueueItem,
	IOrderUpdate,
	IRawOrder,
	ISequenceCacheItem,
	IStringSignedOrder,
	IWsOrderSequenceResponse
} from '../common/types';
import dynamoUtil from '../utils/dynamoUtil';
import orderUtil from '../utils/orderUtil';
import redisUtil from '../utils/redisUtil';
import util from '../utils/util';
import Web3Util from '../utils/Web3Util';

class OrderWatcherServer extends SequenceClient {
	public sequenceMethods = [CST.DB_UPDATE, CST.DB_EXPIRE];
	public orderWatcher: OrderWatcher | null = null;
	public web3Util: Web3Util | null = null;
	public watchingOrders: string[] = [];

	public async handleSequenceResponse(
		res: IWsOrderSequenceResponse,
		cacheItem: ISequenceCacheItem
	) {
		const orderQueueItem: IOrderQueueItem = {
			liveOrder: cacheItem.liveOrder
		};
		const userOrder = await orderUtil.persistOrder(res.method, orderQueueItem);
		if (!userOrder) {
			util.logInfo(`invalid orderHash ${res.orderHash}, ignore`);
			this.removeFromWatch(res.orderHash);
		}
	}

	public async handleOrderWatcherUpdate(pair: string, orderState: OrderState) {
		if (!this.sequenceWsClient) {
			util.logError('sequence service is unavailable');
			return;
		}

		const liveOrder = await orderUtil.getLiveOrderInPersistence(pair, orderState.orderHash);
		if (!liveOrder) {
			util.logInfo(`invalid orderHash ${orderState.orderHash}, ignore`);
			this.removeFromWatch(orderState.orderHash);
			return;
		}

		let method = CST.DB_UPDATE;
		if (orderState.isValid) {
			const remainingAmount = Web3Util.fromWei(
				(orderState as OrderStateValid).orderRelevantState.remainingFillableMakerAssetAmount
			);
			liveOrder.amount = remainingAmount;
		} else {
			const error = (orderState as OrderStateInvalid).error;
			switch (error) {
				case ExchangeContractErrs.OrderCancelExpired:
				case ExchangeContractErrs.OrderFillExpired:
					method = CST.DB_EXPIRE;
					break;
				case ExchangeContractErrs.OrderCancelled:
					method = CST.DB_CANCEL;
					break;
				case ExchangeContractErrs.OrderRemainingFillAmountZero:
					method = CST.DB_FILL;
					break;
				default:
					break;
			}
		}
		const cacheKey = `${method}|${pair}|${orderState.orderHash}`;
		this.requestCache[cacheKey] = {
			pair: pair,
			method: method,
			liveOrder: liveOrder,
			timeout: setTimeout(() => this.handleTimeout(cacheKey), 30000)
		};
		util.logDebug('request added to cache');
		this.requestSequence(method, pair, orderState.orderHash);
	}

	public async addIntoWatch(orderHash: string, signedOrder?: IStringSignedOrder) {
		try {
			if (!signedOrder) {
				const rawOrder: IRawOrder | null = await dynamoUtil.getRawOrder(orderHash);
				if (!rawOrder) {
					util.logDebug('no signed order specified, failed to add');
					return;
				}
				signedOrder = rawOrder.signedOrder as IStringSignedOrder;
			}
			if (this.orderWatcher) {
				await this.orderWatcher.addOrderAsync(orderUtil.parseSignedOrder(signedOrder));
				util.logDebug('succsfully added ' + orderHash);
			}
		} catch (e) {
			util.logDebug('failed to add ' + orderHash + 'error is ' + e);
			this.watchingOrders = this.watchingOrders.filter(hash => hash !== orderHash);
		}
	}

	public async removeFromWatch(orderHash: string) {
		if (!this.watchingOrders.includes(orderHash)) {
			util.logDebug('order is not currently watched');
			return;
		}
		try {
			if (this.orderWatcher) {
				await this.orderWatcher.removeOrder(orderHash);
				util.logDebug('succsfully removed ' + orderHash);
				this.watchingOrders = this.watchingOrders.filter(hash => hash !== orderHash);
			}
		} catch (e) {
			util.logDebug('failed to remove ' + orderHash + 'error is ' + e);
		}
	}

	public async coldStart(pair: string) {
		util.logInfo('start order watcher for ' + pair);
		if (!this.orderWatcher) {
			util.logDebug('orderWatcher is not initiated');
			return;
		}
		const ordersInCache = await redisUtil.hashGetAll(CST.DB_CACHE);
		if (Object.keys(ordersInCache).length) {
			util.logInfo('orders length in DB is ' + Object.keys(ordersInCache).length);
			for (const cacheKey of Object.keys(ordersInCache)) {
				const [method, orderHash] = cacheKey.split('|');
				if (method !== CST.DB_CANCEL && !this.watchingOrders.includes(orderHash)) {
					this.watchingOrders.push(orderHash);
					await this.addIntoWatch(orderHash);
				}
			}
		}

		const liveOrders: ILiveOrder[] = await dynamoUtil.getLiveOrders(pair);
		if (liveOrders.length) {
			util.logInfo('orders length in DB is ' + liveOrders.length);
			for (const order of liveOrders)
				if (!this.watchingOrders.includes(order.orderHash)) {
					this.watchingOrders.push(order.orderHash);
					await this.addIntoWatch(order.orderHash);
				}
		}
	}

	public handleOrderUpdate = (channel: string, orderUpdate: IOrderUpdate) => {
		util.logInfo('receive update from channel: ' + channel);
		const method = orderUpdate.method;
		switch (method) {
			case CST.DB_ADD:
				this.addIntoWatch(orderUpdate.liveOrder.orderHash, orderUpdate.signedOrder);
				break;
			case CST.DB_CANCEL:
				break;
			default:
				break;
		}
	};

	public async startOrderWatcher(web3Util: Web3Util, option: IOption) {
		this.web3Util = web3Util;
		await this.connectToSequenceServer(option.server);
		this.orderWatcher = new OrderWatcher(
			this.web3Util.web3Wrapper.getProvider(),
			option.live ? CST.NETWORK_ID_MAIN : CST.NETWORK_ID_KOVAN
		);
		const pair = option.token + '-' + CST.TOKEN_WETH;

		redisUtil.onOrderUpdate((channel, orderUpdate) =>
			this.handleOrderUpdate(channel, orderUpdate)
		);

		await this.coldStart(pair);
		setInterval(() => this.coldStart(pair), CST.ONE_MINUTE_MS * 60);

		if (option.server) {
			dynamoUtil.updateStatus(pair);
			setInterval(() => dynamoUtil.updateStatus(pair, this.watchingOrders.length), 10000);
		}

		this.orderWatcher.subscribe(async (err, orderState) => {
			if (err || !orderState) {
				util.logError(err ? err : 'orderState empty');
				return;
			}

			this.handleOrderWatcherUpdate(pair, orderState);
		});
	}
}

const orderWatcherServer = new OrderWatcherServer();
export default orderWatcherServer;
