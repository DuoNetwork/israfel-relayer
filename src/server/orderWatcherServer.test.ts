// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';

import { BigNumber, ExchangeContractErrs, OrderState } from '0x.js';
import * as CST from '../common/constants';
import dynamoUtil from '../utils/dynamoUtil';
import orderPersistenceUtil from '../utils/orderPersistenceUtil';
import orderWatcherServer from './orderWatcherServer';

test('remove from watch, not a existing order', async () => {
	const removeOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.orderWatcher = {
		removeOrder: removeOrder
	} as any;
	orderWatcherServer.watchingOrders = {};
	orderWatcherServer.removeFromWatch('orderHash');
	expect(removeOrder).not.toBeCalled();
	expect(orderWatcherServer.watchingOrders).toEqual({});
});

test('remove from watch, exisitng order', async () => {
	const removeOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.orderWatcher = {
		removeOrder: removeOrder
	} as any;

	orderWatcherServer.watchingOrders = { orderHash: {} } as any;
	await orderWatcherServer.removeFromWatch('orderHash');
	expect(removeOrder.mock.calls).toMatchSnapshot();
	expect(orderWatcherServer.watchingOrders).toEqual({});
});

test('updateOrder isValid no userOrder', async () => {
	orderPersistenceUtil.persistOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.updateOrder({
		method: 'method',
		pair: 'pair',
		orderHash: 'orderHash',
		balance: -1
	});
	expect((orderPersistenceUtil.persistOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
});

const userOrder = {
	account: '0xAccount',
	pair: 'pair',
	type: 'type',
	status: 'status',
	orderHash: '0xOrderHash',
	price: 0.123456789,
	amount: 456,
	side: 'side',
	createdAt: 1234560000,
	initialSequence: 1,
	currentSequence: 2,
	updatedBy: 'updatedBy'
};

test('handleOrderWatcherUpdate isValid userOrder', async () => {
	orderPersistenceUtil.persistOrder = jest.fn(() => Promise.resolve(userOrder));
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.updateOrder({
		method: 'method',
		pair: 'pair',
		orderHash: '0xOrderHash',
		balance: -1
	});
	expect((orderPersistenceUtil.persistOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(orderWatcherServer.removeFromWatch as jest.Mock).not.toBeCalled();
});

const signedOrder = {
	senderAddress: 'senderAddress',
	makerAddress: 'makerAddress',
	takerAddress: 'takerAddress',
	makerFee: '0',
	takerFee: '0',
	makerAssetAmount: '123',
	takerAssetAmount: '456',
	makerAssetData: 'makerAssetData',
	takerAssetData: 'takerAssetData',
	salt: '789',
	exchangeAddress: 'exchangeAddress',
	feeRecipientAddress: 'feeRecipientAddress',
	expirationTimeSeconds: '1234567890',
	signature: 'signature'
};

test('addIntoWatch with signed order fillable', async () => {
	const addOrderAsync = jest.fn(() => Promise.resolve());
	orderWatcherServer.orderWatcher = {
		addOrderAsync: addOrderAsync
	} as any;
	dynamoUtil.getRawOrder = jest.fn(() => Promise.resolve({}));
	orderWatcherServer.watchingOrders = {};
	orderWatcherServer.web3Util = {
		validateOrderFillable: jest.fn(() => Promise.resolve(true))
	} as any;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	await orderWatcherServer.addIntoWatch('orderHash', signedOrder);
	expect(dynamoUtil.getRawOrder as jest.Mock).not.toBeCalled();
	expect(orderWatcherServer.updateOrder as jest.Mock).not.toBeCalled();
	expect(addOrderAsync.mock.calls).toMatchSnapshot();
	expect(orderWatcherServer.watchingOrders).toMatchSnapshot();
});

test('addIntoWatch with signed order non fillable', async () => {
	const addOrderAsync = jest.fn(() => Promise.resolve());
	orderWatcherServer.orderWatcher = {
		addOrderAsync: addOrderAsync
	} as any;
	dynamoUtil.getRawOrder = jest.fn(() => Promise.resolve({}));
	orderWatcherServer.watchingOrders = {};
	orderWatcherServer.web3Util = {
		validateOrderFillable: jest.fn(() => Promise.resolve(false))
	} as any;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	await orderWatcherServer.addIntoWatch('orderHash', signedOrder);
	expect(dynamoUtil.getRawOrder as jest.Mock).not.toBeCalled();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(addOrderAsync.mock.calls).toMatchSnapshot();
	expect(orderWatcherServer.watchingOrders).toMatchSnapshot();
});

test('addIntoWatch no signed order fillable', async () => {
	const addOrderAsync = jest.fn(() => Promise.resolve());
	orderWatcherServer.orderWatcher = {
		addOrderAsync: addOrderAsync
	} as any;
	orderWatcherServer.watchingOrders = {};
	orderWatcherServer.web3Util = {
		validateOrderFillable: jest.fn(() => Promise.resolve(true))
	} as any;
	dynamoUtil.getRawOrder = jest.fn(() =>
		Promise.resolve({
			orderHash: 'orderHash',
			signedOrder: signedOrder
		})
	);
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	await orderWatcherServer.addIntoWatch('orderHash');
	expect(addOrderAsync.mock.calls).toMatchSnapshot();
	expect((dynamoUtil.getRawOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(orderWatcherServer.updateOrder as jest.Mock).not.toBeCalled();
	expect(orderWatcherServer.watchingOrders).toMatchSnapshot();
});

test('addIntoWatch no signed order and no rawOrder', async () => {
	const addOrderAsync = jest.fn(() => Promise.resolve());
	orderWatcherServer.orderWatcher = {
		addOrderAsync: addOrderAsync
	} as any;
	orderWatcherServer.watchingOrders = {};
	orderWatcherServer.web3Util = {
		validateOrderFillable: jest.fn(() => Promise.resolve(true))
	} as any;

	dynamoUtil.getRawOrder = jest.fn(() => Promise.resolve());
	await orderWatcherServer.addIntoWatch('orderHash');
	expect(addOrderAsync).not.toBeCalled();
	expect((orderWatcherServer.web3Util as any).validateOrderFillable).not.toBeCalled();
	expect(orderWatcherServer.watchingOrders).toEqual({});
});

const orderPersistRequest = {
	method: 'method',
	pair: 'pair',
	orderHash: 'orderHash',
	balance: 456,
	signedOrder: signedOrder
};
test('handle orderUpdate invalid method', async () => {
	orderWatcherServer.orderWatcher = null;
	orderWatcherServer.addIntoWatch = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderUpdate('channel', orderPersistRequest);
	expect(orderWatcherServer.addIntoWatch as jest.Mock).not.toBeCalled();
	expect(orderWatcherServer.removeFromWatch as jest.Mock).not.toBeCalled();
});

test('handle orderUpdate add', async () => {
	orderWatcherServer.orderWatcher = null;
	orderWatcherServer.addIntoWatch = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	orderPersistRequest.method = CST.DB_ADD;
	await orderWatcherServer.handleOrderUpdate('channel', orderPersistRequest);
	expect((orderWatcherServer.addIntoWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect(orderWatcherServer.removeFromWatch as jest.Mock).not.toBeCalled();
});

test('handle orderUpdate terminate', async () => {
	orderWatcherServer.orderWatcher = null;
	orderWatcherServer.addIntoWatch = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	orderPersistRequest.method = CST.DB_TERMINATE;
	await orderWatcherServer.handleOrderUpdate('channel', orderPersistRequest);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect(orderWatcherServer.addIntoWatch as jest.Mock).not.toBeCalled();
});

const orderStateValid: OrderState = {
	isValid: true,
	orderHash: 'orderHash',
	orderRelevantState: {
		remainingFillableMakerAssetAmount: new BigNumber(567)
	} as any
};

test('handleOrderWatcherUpdate isValid', async () => {
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateValid);
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
});

const orderStateInValid: OrderState = {
	isValid: false,
	orderHash: 'orderHash',
	error: ExchangeContractErrs.OrderCancelExpired
};

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderCancelExpired', async () => {
	orderStateInValid.error = ExchangeContractErrs.OrderCancelExpired;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderFillExpired', async () => {
	orderStateInValid.error = ExchangeContractErrs.OrderFillExpired;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderCancelled', async () => {
	orderStateInValid.error = ExchangeContractErrs.OrderCancelled;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderRemainingFillAmountZero', async () => {
	orderStateInValid.error = ExchangeContractErrs.OrderRemainingFillAmountZero;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderRemainingFillAmountZero', async () => {
	orderStateInValid.error = ExchangeContractErrs.OrderRemainingFillAmountZero;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerBalance', async () => {
	orderStateInValid.error = ExchangeContractErrs.InsufficientTakerBalance;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerAllowance', async () => {
	orderStateInValid.error = ExchangeContractErrs.InsufficientTakerAllowance;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerFeeBalance', async () => {
	orderStateInValid.error = ExchangeContractErrs.InsufficientTakerFeeBalance;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerFeeAllowance', async () => {
	orderStateInValid.error = ExchangeContractErrs.InsufficientTakerFeeAllowance;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientMakerFeeBalance', async () => {
	orderStateInValid.error = ExchangeContractErrs.InsufficientMakerFeeBalance;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientMakerFeeAllowance', async () => {
	orderStateInValid.error = ExchangeContractErrs.InsufficientMakerFeeAllowance;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect((orderWatcherServer.removeFromWatch as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('handleOrderWatcherUpdate inValid default', async () => {
	orderStateInValid.error = '' as any;
	orderWatcherServer.updateOrder = jest.fn(() => Promise.resolve());
	orderWatcherServer.removeFromWatch = jest.fn(() => Promise.resolve());
	await orderWatcherServer.handleOrderWatcherUpdate(orderStateInValid);
	expect(orderWatcherServer.removeFromWatch as jest.Mock).not.toBeCalled();
	expect((orderWatcherServer.updateOrder as jest.Mock).mock.calls).toMatchSnapshot();
});
