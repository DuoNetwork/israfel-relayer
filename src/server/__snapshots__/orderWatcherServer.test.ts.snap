// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addIntoWatch no signed order fillable 1`] = `
Array [
  Array [
    Object {
      "exchangeAddress": "exchangeAddress",
      "expirationTimeSeconds": "1234567890",
      "feeRecipientAddress": "feeRecipientAddress",
      "makerAddress": "makerAddress",
      "makerAssetAmount": "123",
      "makerAssetData": "makerAssetData",
      "makerFee": "0",
      "salt": "789",
      "senderAddress": "senderAddress",
      "signature": "signature",
      "takerAddress": "takerAddress",
      "takerAssetAmount": "456",
      "takerAssetData": "takerAssetData",
      "takerFee": "0",
    },
  ],
]
`;

exports[`addIntoWatch no signed order fillable 2`] = `
Array [
  Array [
    "orderHash",
  ],
]
`;

exports[`addIntoWatch no signed order fillable 3`] = `
Object {
  "orderHash": Object {
    "exchangeAddress": "exchangeAddress",
    "expirationTimeSeconds": "1234567890",
    "feeRecipientAddress": "feeRecipientAddress",
    "makerAddress": "makerAddress",
    "makerAssetAmount": "123",
    "makerAssetData": "makerAssetData",
    "makerFee": "0",
    "salt": "789",
    "senderAddress": "senderAddress",
    "signature": "signature",
    "takerAddress": "takerAddress",
    "takerAssetAmount": "456",
    "takerAssetData": "takerAssetData",
    "takerFee": "0",
  },
}
`;

exports[`addIntoWatch with signed order fillable 1`] = `
Array [
  Array [
    Object {
      "exchangeAddress": "exchangeAddress",
      "expirationTimeSeconds": "1234567890",
      "feeRecipientAddress": "feeRecipientAddress",
      "makerAddress": "makerAddress",
      "makerAssetAmount": "123",
      "makerAssetData": "makerAssetData",
      "makerFee": "0",
      "salt": "789",
      "senderAddress": "senderAddress",
      "signature": "signature",
      "takerAddress": "takerAddress",
      "takerAssetAmount": "456",
      "takerAssetData": "takerAssetData",
      "takerFee": "0",
    },
  ],
]
`;

exports[`addIntoWatch with signed order fillable 2`] = `
Object {
  "orderHash": Object {
    "exchangeAddress": "exchangeAddress",
    "expirationTimeSeconds": "1234567890",
    "feeRecipientAddress": "feeRecipientAddress",
    "makerAddress": "makerAddress",
    "makerAssetAmount": "123",
    "makerAssetData": "makerAssetData",
    "makerFee": "0",
    "salt": "789",
    "senderAddress": "senderAddress",
    "signature": "signature",
    "takerAddress": "takerAddress",
    "takerAssetAmount": "456",
    "takerAssetData": "takerAssetData",
    "takerFee": "0",
  },
}
`;

exports[`addIntoWatch with signed order non fillable 1`] = `
Array [
  Array [
    Object {
      "balance": 0,
      "method": "update",
      "orderHash": "orderHash",
      "pair": "pair",
    },
  ],
]
`;

exports[`addIntoWatch with signed order non fillable 2`] = `
Array [
  Array [
    Object {
      "exchangeAddress": "exchangeAddress",
      "expirationTimeSeconds": "1234567890",
      "feeRecipientAddress": "feeRecipientAddress",
      "makerAddress": "makerAddress",
      "makerAssetAmount": "123",
      "makerAssetData": "makerAssetData",
      "makerFee": "0",
      "salt": "789",
      "senderAddress": "senderAddress",
      "signature": "signature",
      "takerAddress": "takerAddress",
      "takerAssetAmount": "456",
      "takerAssetData": "takerAssetData",
      "takerFee": "0",
    },
  ],
]
`;

exports[`addIntoWatch with signed order non fillable 3`] = `
Object {
  "orderHash": Object {
    "exchangeAddress": "exchangeAddress",
    "expirationTimeSeconds": "1234567890",
    "feeRecipientAddress": "feeRecipientAddress",
    "makerAddress": "makerAddress",
    "makerAssetAmount": "123",
    "makerAssetData": "makerAssetData",
    "makerFee": "0",
    "salt": "789",
    "senderAddress": "senderAddress",
    "signature": "signature",
    "takerAddress": "takerAddress",
    "takerAssetAmount": "456",
    "takerAssetData": "takerAssetData",
    "takerFee": "0",
  },
}
`;

exports[`handle orderUpdate add 1`] = `
Array [
  Array [
    "orderHash",
    Object {
      "exchangeAddress": "exchangeAddress",
      "expirationTimeSeconds": "1234567890",
      "feeRecipientAddress": "feeRecipientAddress",
      "makerAddress": "makerAddress",
      "makerAssetAmount": "123",
      "makerAssetData": "makerAssetData",
      "makerFee": "0",
      "salt": "789",
      "senderAddress": "senderAddress",
      "signature": "signature",
      "takerAddress": "takerAddress",
      "takerAssetAmount": "456",
      "takerAssetData": "takerAssetData",
      "takerFee": "0",
    },
  ],
]
`;

exports[`handle orderUpdate terminate 1`] = `
Array [
  Array [
    "orderHash",
  ],
]
`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientMakerFeeAllowance 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientMakerFeeAllowance 2`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientMakerFeeBalance 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientMakerFeeBalance 2`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerAllowance 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerAllowance 2`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerBalance 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerBalance 2`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerFeeAllowance 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerFeeAllowance 2`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerFeeBalance 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.InsufficientTakerFeeBalance 2`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderCancelExpired 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderCancelExpired 2`] = `
Array [
  Array [
    Object {
      "balance": -1,
      "method": "terminate",
      "orderHash": "orderHash",
      "pair": "pair",
    },
  ],
]
`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderCancelled 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderCancelled 2`] = `
Array [
  Array [
    Object {
      "balance": -1,
      "method": "terminate",
      "orderHash": "orderHash",
      "pair": "pair",
    },
  ],
]
`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderFillExpired 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderFillExpired 2`] = `
Array [
  Array [
    Object {
      "balance": -1,
      "method": "terminate",
      "orderHash": "orderHash",
      "pair": "pair",
    },
  ],
]
`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderRemainingFillAmountZero 1`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderRemainingFillAmountZero 2`] = `
Array [
  Array [
    Object {
      "balance": 0,
      "method": "terminate",
      "orderHash": "orderHash",
      "pair": "pair",
    },
  ],
]
`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderRemainingFillAmountZero 3`] = `Array []`;

exports[`handleOrderWatcherUpdate inValid ExchangeContractErrs.OrderRemainingFillAmountZero 4`] = `
Array [
  Array [
    Object {
      "balance": 0,
      "method": "terminate",
      "orderHash": "orderHash",
      "pair": "pair",
    },
  ],
]
`;

exports[`handleOrderWatcherUpdate inValid default 1`] = `Array []`;

exports[`handleOrderWatcherUpdate isValid 1`] = `
Array [
  Array [
    Object {
      "balance": 5.67e-16,
      "method": "update",
      "orderHash": "orderHash",
      "pair": "pair",
    },
  ],
]
`;

exports[`handleOrderWatcherUpdate isValid 2`] = `Array []`;

exports[`handleOrderWatcherUpdate isValid userOrder 1`] = `
Array [
  Array [
    Object {
      "balance": -1,
      "method": "method",
      "orderHash": "0xOrderHash",
      "pair": "pair",
    },
    false,
  ],
]
`;

exports[`remove from watch, exisitng order 1`] = `
Array [
  Array [
    "orderHash",
  ],
]
`;

exports[`updateOrder isValid no userOrder 1`] = `
Array [
  Array [
    Object {
      "balance": -1,
      "method": "method",
      "orderHash": "orderHash",
      "pair": "pair",
    },
    false,
  ],
]
`;

exports[`updateOrder isValid no userOrder 2`] = `
Array [
  Array [
    "orderHash",
  ],
]
`;
